src/hti/re_dash.cljd

```diff
@@ -349,65 +350,134 @@
 
 ;; Flows
 
-(defonce ^:private flows (atom []))
+(defonce ^:private flows (atom {:definitions {}
+                                :sorted-ids  []}))
+
+(def ^:private flow? map?)
+
+(defn- get-flow
+  [flow-id definitions]
+  (or (flow-id definitions)
+      (throw (Exception. (str "Flow not found: " flow-id)))))
+
+(defn- derived-input-path
+  "If input-path is a ref to another flow, return its path"
+  [definitions input-path]
+  (when (flow? input-path)
+    (-> input-path
+        ::flow<-
+        (get-flow definitions)
+        :path)))
 
 (defn- resolve-inputs
-  [db inputs]
+  "Resolves the flow inputs to their current values in app-db"
+  [db inputs definitions]
   (->> (map (fn [[input-key input-path]]
-              {input-key (get-in db input-path)})
+              {input-key (get-in db
+                                 (or (derived-input-path definitions input-path)
+                                     input-path))})
             inputs)
        (into {})))
 
 
-(def flow-interceptor
+(def ^:private flow-interceptor
   (interceptors/->interceptor
     :re-dash/flow
     (fn [{{db :db} :effects
           :as      context}]
-      (assoc-in context [:effects :db]
-                (reduce (fn [new-db {:keys [inputs output path]}]
-                          (let [new-db-resolved-inputs (resolve-inputs new-db inputs)
-                                old-db-resolved-inputs (resolve-inputs @app-db inputs)]
-                            (if (not= new-db-resolved-inputs old-db-resolved-inputs)
-                              (assoc-in new-db path (output new-db-resolved-inputs))
-                              new-db)))
-                        db
-                        @flows)))
+      (let [app-db                           @app-db
+            {:keys [definitions sorted-ids]} @flows]
+        (assoc-in context [:effects :db]
+                  (reduce (fn [new-db flow-id]
+                            (let [{:keys [inputs output path]}
+                                  (get-flow flow-id definitions)
+
+                                  new-db-resolved-inputs
+                                  (resolve-inputs new-db inputs definitions)
+
+                                  old-db-resolved-inputs
+                                  (resolve-inputs app-db inputs definitions)]
+
+                              (if (not= new-db-resolved-inputs old-db-resolved-inputs)
+                                (assoc-in new-db path (output new-db-resolved-inputs))
+                                new-db)))
+                          db
+                          sorted-ids))))
     nil))
 
-(defn- flows-match?
-  [flow1 flow2]
-  #(= (:id flow1) (:id flow2)))
 
-(defn- sort-flows
-  [new-flows]
+(defn flow<-
+  "Return a representation of a referenced input flow"
+  [flow-id]
+  {::flow<- flow-id})
+
+(defn- sort-flow-ids
+  "Sort all Flow definition dependency graphs
```

Comment:
all? do we have to worry about cyclic dependency between flows that aren't live?